src/System.Buffers/src/System/Buffers/BufferInternal.cs

```diff
@@ -192,5 +192,121 @@ public static unsafe void MemoryCopy(void* source, void* destination, int destin
 
             Memmove((byte*)destination, (byte*)source, checked((uint)sourceBytesToCopy));
         }
+        
+        // TODO: This is naive implementation. Do we have any better native implementation?
+        // TODO: Should we start thinking about supporting long lengths?
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe bool MemoryEqual(void* lhs, int lhsLength, void* rhs, int rhsLength)
+        {
+            // 1. Fast paths
+            if (lhsLength != rhsLength)
+                return false;
+            if (lhs == rhs)
+                return true;
+            if (lhsLength == 0)
+                return true;
+            if (lhsLength == 4)
+                return *(UInt32*)lhs == *(UInt32*)rhs;
+            if (lhsLength == 8)
+                return *(UInt64*)lhs == *(UInt64*)rhs;
+            
+            switch (lhsLength)
+            {
+                case 1: return *(byte*)lhs == *(byte*)rhs;
+                case 2: return *(UInt16*)lhs == *(UInt16*)rhs;
+                case 3: return *(UInt16*)lhs == *(UInt16*)rhs &&  *((byte*)lhs + 2) == *((byte*)rhs + 2);
+            }
+            
+            // 2. We split bytes into two groups of: (n/4)*4 and n%4 bytes
+            
+            // TODO: should we play with alignment here?
+            UInt32* a = (UInt32*)lhs;
+            UInt32* b = (UInt32*)rhs;
+            
+            // I think those are one processor instruction, so keeping it together
+            int count = lhsLength / sizeof(UInt32);
+            int remainder = lhsLength % sizeof(UInt32);
```

Comment:
Yeah, I checked our JITs some time ago and they generate two idiv instructions. @CarolEidt was thinking about a way to add an intrinsic for this or pattern optimizations.